src/transforms/javascript.rs

```diff
@@ -0,0 +1,789 @@
+        let source = match (source, path) {
+            (Some(source), None) => source,
+            (None, Some(path)) => fs::read_to_string(&path).map_err(|err| {
+                format!("Cannot load JavaScript source from \"{}\": {}", path, err)
+            })?,
+            (Some(_), Some(_)) => {
+                return Err("\"source\" and \"path\" cannot be provided together".to_string())
+            }
+            (None, None) => {
+                return Err("Either \"source\" or \"path\" should be provided".to_string())
+            }
+        };
+
+        // validate handler parameter if present
+        if let Some(ref handler) = handler {
+            if !JS_VALID_IDENTIFIER.is_match(&handler) {
+                return Err(format!(
+                    "Handler name \"{}\" is not a valid JavaScript identifier",
+                    handler
+                ));
+            }
+            if JS_RESERVED_KEYWORDS.contains(&handler[..]) {
+                return Err(format!(
+                    "Handler name \"{}\" is reserved in JavaScript and cannot be used",
+                    handler
+                ));
+            }
+        }
+
+        // init QuickJS context
+        let mut builder = Context::builder();
+        if let Some(memory_limit) = memory_limit {
+            builder = builder.memory_limit(memory_limit);
+        }
+        let ctx = builder
+            .build()
+            .map_err(|err| format!("Cannot create JavaScript runtime: {}", err))?;
+
+        // inject handler source
+        let source = if let Some(handler) = handler {
+            format!("{}; __vector_handler = {}", source, handler)
+        } else {
+            format!(r"__vector_handler = ({})", source)
+        };
+        ctx.eval(&source)
+            .map_err(|err| format!("Cannot create handler: {}", err))?;
+
+        // check that handler is a function
+        let handler_is_a_function: bool = ctx
+            .eval_as(r"typeof __vector_handler === 'function'")
+            .map_err(|err| format!("Cannot validate handler: {}", err))?;
+        if !handler_is_a_function {
+            return Err("Handler is not a function".to_string());
+        }
+
+        Ok(Self { ctx })
+    }
+
+    pub fn process(&self, output: &mut Vec<Event>, event: Event) -> Result<(), String> {
+        let encoded = encode(event)?;
+        let transformed = self
+            .ctx
+            .call_function("__vector_handler", vec![encoded])
+            .map_err(|err| format!("Runtime error in JavaScript code: {}", err))?;
+        write_value(transformed, output)
+    }
+}
+
+fn encode(event: Event) -> Result<JsValue, String> {
+    let js_event = event
+        .as_log()
+        .all_fields()
+        .map(|(key, value)| {
+            (
+                key.to_string(),
+                match value {
+                    ValueKind::Bytes(v) => {
+                        JsValue::String(std::str::from_utf8(v).unwrap().to_string())
```

Comment:
Actually I'm not sure what would be the best way to deal with invalid UTF-8 sequences here. One possibility is to use lossy and another one is to convert that sequences to [`ArrayBuffer`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) instead of `String` (needs to be added to `quick-js` crate first).

Reply:
If it is used in other places, then lossy conversion is fine.

I was in doubt because I was thinking about someone sending binary data (for example, MessagePack) to Vector and then wanting to simultaneously store it to S3 and create metrics data from it. In that case the message would not be valid UTF-8 and lossy conversion cannot be used.

Reply:
> for example, MessagePack)
👎 MessagePack. Won't it be normalized into our event structure by the time it reaches this transform?

PS - I'm not a huge fan of MessagePack and plan to distill my thoughts into a post at some point in the future.

Reply:
Yeah, let's do lossy conversion its what we have defaulted too.

Reply:
> 👎 MessagePack. Won't it be normalized into our event structure by the time it reaches this transform?

I mentioned MessagePack as an arbitrary example of a binary format that is not supported by Vector, but which might be utilized by user's applications to serialize the data that they, for example, write to some log file, and which are then ingested by Vector using `file` source. The produced events would contain opaque blobs in the `message` field in that case, which are not valid UTF-8.

Reply:
Got it. Your concern is correct. It'd be interesting to think about that when we work on #534.